### PR TITLE
PRJ-199 Support native double buffering

### PR DESCRIPTION
**Pull request by SerVB**
This PR fixes PRJ-199 (flickering) and remaining problems with PRJ-20 (scrolling glitches).

I will rebase and request reviews after #100 is merged.

Commits here can be reviewed separately.

Server-side changes: https://github.com/JetBrains/projector-server/pull/100

### CODE REVIEW

**Review by ARTI1208 — COMMENTED** [review]
Is there a good reproducer for PRJ-199? Currently in git log I see no difference between this branch and master. But what I can see is that with these changes graphs are now practically working without my hack with transformer (however there are still some artifacts without the hack, but not so severe as before)

Also, with this PR I see many errors like
`java.lang.Error: no ComponentUI class for: javax.swing.JProgressBar[,0,0,0x0,invalid,alignmentX=0.0,alignmentY=0.0,border=,flags=0,maximumSize=,minimumSize=,preferredSize=,orientation=HORIZONTAL,paintBorder=false,paintString=false,progressString=,indeterminateString=false]`
in logs, but everything seems to work. Could you investigate where it comes from and can it break anything?

As I can't reproduce the problem PR supposed to fix and due to those errors I won't approve for now. Please, clarify those two things (or at least the first one)

File: projector-server-core/src/main/kotlin/org/jetbrains/projector/server/core/util/distinctUpdatedOnscreenSurfaces.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.jetbrains.projector.server.core.util
+
+import org.jetbrains.projector.common.protocol.toClient.ServerDrawCommandsEvent
+import org.jetbrains.projector.common.protocol.toClient.ServerEvent
+import org.jetbrains.projector.common.protocol.toClient.ServerWindowPaintEvent
+
+public fun List<ServerEvent>.distinctUpdatedOnscreenSurfaces(): List<ServerDrawCommandsEvent.Target.Onscreen> = this
```
[the remark this answers is not in the document]
Comment (SerVB) [reply projector-server-core/src/main/kotlin/org/jetbrains/projector/server/core/util/distinctUpdatedOnscreenSurfaces.kt:30]: The latter one

**Comment by SerVB** [comment]
Sure!

> Is there a good reproducer for PRJ-199? Currently in git log I see no difference between this branch and master.

Glitchy scrolling in Git Log is almost solved by PRJ-20 PR, so indeed it's look similar here and in master.

This PR fixes these tickets and they are easy to reproduce, please try:
https://youtrack.jetbrains.com/issue/PRJ-622
https://youtrack.jetbrains.com/issue/PRJ-527
https://youtrack.jetbrains.com/issue/PRJ-774

I've sent you the video with the first two in Slack.

> But what I can see is that with these changes graphs are now practically working without my hack with transformer (however there are still some artifacts without the hack, but not so severe as before)

I've looked into it and there can be still severe glitches if we allow Swing Double Buffering, so I suggest to leave that transformer hack.

> with this PRs I see many errors like java.lang.Error: no ComponentUI class for

Good catch, thank you very much for noticing that. It's a regression of the last optimization, without it there is no such problem logged. I'll recheck tomorrow, probably some actions are performed in an incorrect thread (I had quite the same issue before when preparing this PR without optimizations and the solution was to carefully move some long actions out of the EDT).

**Comment by ARTI1208** [comment]
> This PR fixes these tickets
> I've sent you the video with the first two in Slack
Thanks, I see these two first tickets are now fixed (but magnifier still blinks)

> I suggest to leave that transformer hack
That's for sure, I haven't proposed to remove it, just noticed that situation got better

**Comment by SerVB** [comment]
Found the problem, please see the commit https://github.com/JetBrains/projector-server/pull/100/commits/b2bc1d02c6c2f7cf0aaa1c0a89e707176f56045d. It's an interesting one. The only reason I've managed to find it is because I already bumped into it a year ago or so: EDT shouldn't be used too early by us.

Regarding diagrams, don't blink for me as they were before the PR. The performance could be improved, but I think it's OK for now. I haven't managed to run diagrams on the previous commit (it was an exception logged), maybe blinking was presented there. On the latest commit, there is no blinking:

https://user-images.githubusercontent.com/20105593/144263740-894954db-7829-4b79-bef6-aee07998a0af.mp4

Please recheck.

**Review by ARTI1208 — COMMENTED** [review]
Exceptions are now indeed gone.
Speaking about magnifier - may be I used a wrong word for its behavior. I meant that it stops work when you drag and start again only when drag finishes (without Projector it works throughout the process). I thought the PR was showing just this (and your screencast proves it, if you held ALT during dragging). But it looks like a different problem, you agree? If yes - I'll approve

**Comment by ARTI1208** [comment]
Oh. and thanks for note about using EDT too early - I have the same errors with jdk17 now, may it'll help..

**Comment by SerVB** [comment]
> I thought the PR was showing just this (and your screencast proves it, if you held ALT during dragging). But it looks like a different problem, you agree?

Now I see it. Indeed I held ALT during dragging but the magnifier works only when I stop moving.

Let me file a separate ticket for it.

**Comment by SerVB** [comment]
Here it is: https://youtrack.jetbrains.com/issue/PRJ-790

**Review by ARTI1208 — APPROVED** [review]

**Pull request merged**